sockeye/optimizers.py

```diff
@@ -0,0 +1,202 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not
+# use this file except in compliance with the License. A copy of the License
+# is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+Extra optimizers not included in MXNet.
+"""
+
+from abc import abstractmethod
+from collections import namedtuple
+import math
+
+from mxnet.ndarray import NDArray, sqrt, zeros
+from mxnet.optimizer import Optimizer
+
+from sockeye.utils import check_condition
+
+BatchState = namedtuple("BatchState", ["metric_val"])
+CheckpointState = namedtuple("CheckpointState", ["checkpoint", "metric_val"])
+
+
+class SockeyeOptimizer(Optimizer):
+    """
+    Optimizer that has access to additional information from the last batch and the last chekpoint
+    when updating weights.
+    """
+    def __init__(self, **kwargs):
+        self.batch_state = None
+        self.checkpoint_state = None
+        super().__init__(**kwargs)
+
+    def pre_update_batch(self, batch_state: BatchState):
+        """
+        Called automatically prior to `update()` for each batch.
+        """
+        self.batch_state = batch_state
+
+    def pre_update_checkpoint(self, checkpoint_state: CheckpointState):
+        """
+        Called automatically at each checkpoint.
+        """
+        self.checkpoint_state = checkpoint_state
+
+    @abstractmethod
+    def update(self, index, weight, grad, state):
+        """
+        Called automatically as normal.
+        """
+        pass
+
+
+# convenience wrapper for Optimizer.Register
+register = Optimizer.register   # pylint: disable=invalid-name
+
+
+@register
+class Eve(SockeyeOptimizer):
+    """
+    The Eve optimizer is an extended version of Adam that incorporates feedback from the objective
+    function to further adapt the learning rate.
+        * "Improving Stochastic Gradient Descent with Feedback"
+          Jayanth Koushik; Hiroaki Hayashi (https://arxiv.org/abs/1611.01505)
+
+    This version allows using validation checkpoint loss in addition to training batch loss.
+
+    Eve does not currently support rescaling gradients, clipping gradients, or weight decay.
```

Comment:
lets log a warning about this in the Eve constructor, i.e. if rescale_gradient etc is present in kwargs.

Also, it seems like Eve is generally not supporting l1/l2 regularizations (Section 3.1), so using weight decay should probably throw an error to indicate this.

Reply:
Good idea to warn early about that.  I'll add some `check_condition` statements.